7th commit:

waspp01 opened the pull request:
- Created Category entity with respective controller
- Created repository and service layers for Category
- Created DTOs for Category
- Created Mapper for Category
- Changed Book entity according to a Category implementation
- Changed BookService layer according to a Category implementation
- Added Liquibase implementations for new tables - categories and books_categories
- Added pagination, sorting and swagger support
- Other minor changes

waspp01 committed:
7th commit:
- Created Category entity with respective controller
- Created repository and service layers for Category
- Created DTOs for Category
- Created Mapper for Category
- Changed Book entity according to a Category implementation
- Changed BookService layer according to a Category implementation
- Added Liquibase implementations for new tables - categories and books_categories
- Added pagination, sorting and swagger support
- Other minor changes

Rommelua commented on src/main/java/com/example/onlinebookstore/controller/BookController.java:
```diff
@@ -30,45 +33,55 @@ public class BookController {
     @PostMapping
     @Operation(summary = "Create a new book", description = "Create a new book")
-    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
+    public BookDto createBook(Authentication authentication,
```
This is 'Add Category model' task. Do not add extra staff. And we do not need Authentication here

Rommelua commented on src/main/java/com/example/onlinebookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,91 @@
+    @Operation(summary = "Get all categories", description = "Get a list of all categories")
+    @ApiResponse(responseCode = "200", description = "All categories",
+            content = {@Content(mediaType = "application/json")})
+    public List<CategoryDto> getAll(Authentication authentication, Pageable pageable) {
```
```suggestion
    public List<CategoryDto> getAll(Pageable pageable) {
```

Rommelua commented on src/main/java/com/example/onlinebookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,91 @@
+    @ApiResponse(responseCode = "200", description = "All categories",
+            content = {@Content(mediaType = "application/json")})
+    public List<CategoryDto> getAll(Authentication authentication, Pageable pageable) {
+        return categoryService.findAll();
```
use pageable

Rommelua replied:
not fixed

Rommelua commented on src/main/java/com/example/onlinebookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,91 @@
+    public List<BookDtoWithoutCategoryIds> getBookByCategory(Authentication authentication,
+                                                             Pageable pageable,
+                                                             @PathVariable Long id) {
+        return bookService.findAllByCategoryId(id);
```
use pageable

Rommelua commented on src/main/java/com/example/onlinebookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,91 @@
+    @Operation(summary = "Get books by category", description = "Books by category")
+    @ApiResponse(responseCode = "200", description = "Books by category",
+            content = {@Content(mediaType = "application/json")})
+    public List<BookDtoWithoutCategoryIds> getBookByCategory(Authentication authentication,
```
```suggestion
    public List<BookDtoWithoutCategoryIds> getBookByCategory(
```

Rommelua commented on src/main/java/com/example/onlinebookstore/model/Category.java:
```diff
@@ -0,0 +1,31 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotNull
```
@Column(nullable = false)

Rommelua commented on src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java:
```diff
@@ -21,11 +25,14 @@ public class BookServiceImpl implements BookService {
     public BookDto save(CreateBookRequestDto requestDto) {
-        return bookMapper.toDto(
-                bookRepository.save(bookMapper.toModel(requestDto)));
+        Set<Category> categories = categoryRepository.findByIdIn(requestDto.getCategoryIds());
+        Book book = bookMapper.toModel(requestDto);
+        book.setCategories(categories);
```
```suggestion
        Book book = bookMapper.toModel(requestDto);
```
add to mapper aftermapping method. We do not need to fetch categories from DB - just create objects using Ids

Rommelua commented on src/main/java/com/example/onlinebookstore/repository/category/CategoryRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    public Set<Category> findByIdIn(Set<Long> categories);
```
```suggestion
```

Rommelua commented on src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:
```diff
@@ -34,7 +34,11 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
         user.setLastName(requestDto.getLastName());
         user.setShippingAddress(requestDto.getShippingAddress());
-        user.setRoles(Set.of(roleRepository.findRoleByName(Role.RoleName.USER)));
+        if (requestDto.getEmail().equals("admin@i.ua")) {
+            user.setRoles(Set.of(roleRepository.findRoleByName(Role.RoleName.ADMIN)));
+        } else {
+            user.setRoles(Set.of(roleRepository.findRoleByName(Role.RoleName.USER)));
+        }
```
remove this hardcode)

waspp01 committed:
7th commit:
- Deleted redundant Authorizations from Book and Category controllers
- Deleted findByIdIn method from CategoryRepository. Deleted its usage in service layer. Implemented @AfterMapping instead
- Deleted redundant hardcode

Rommelua commented on src/main/java/com/example/onlinebookstore/model/Book.java:
```diff
@@ -39,4 +45,10 @@ public class Book {
+    @JoinTable(name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id"))
+    @ManyToMany(fetch = FetchType.EAGER)
```
do not use EAGER

waspp01 committed:
7th commit:
- Added pageable usage where needed
- Changed the name of findAllByCategoryId methid in CategoryController
- Fixed setCategories in BookMapper
- Fixed Eager initialization
- Added new methods into BookRepository for Lazy initialization
- Added new methods into UserRepository for Lazy initialization
- Changed the mane of findAllByCategoryId

sokimaaa commented on src/main/java/com/example/onlinebookstore/controller/CategoryController.java:
```diff
@@ -0,0 +1,86 @@
+    @ApiResponse(responseCode = "200", description = "Update a category",
+            content = {@Content(mediaType = "application/json")})
+    public CategoryDto updateCategory(@PathVariable Long id,
+            @RequestBody CreateCategoryRequestDto categoryRequestDto) {
```
should be validated user input, e.g. @Valid.

sokimaaa commented on src/main/java/com/example/onlinebookstore/model/Category.java:
```diff
@@ -0,0 +1,37 @@
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    public Category(Long id) {
+        this.id = id;
+    }
```
Do you use this constructor? if not it should be removed

waspp01 replied:
It is used here:
@AfterMapping
    default void setCategories(@MappingTarget Book book, CreateBookRequestDto bookDto) {
        book.setCategories(bookDto
                .getCategoryIds()
                .stream()
                .map(Category::new)
                .collect(Collectors.toSet()));
    }

sokimaaa commented on src/main/java/com/example/onlinebookstore/model/Category.java:
```diff
@@ -0,0 +1,37 @@
+        this.id = id;
+    }
+
+    public Category() {
+    }
```
add @NoArgsConstructor annotation from lombok instead of writing constructor manually

sokimaaa commented on src/main/java/com/example/onlinebookstore/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public CategoryDto getById(Long id) {
+        Category category = categoryRepository.findById(id).orElseThrow(
+                () -> new EntityNotFoundException("Can't find category by id " + id));
+        return categoryMapper.toDto(category);
```
use Optional#map to reduce code.
```suggestion
        return categoryRepository.findById(id).map(categoryMapper::toDto).orElseThrow(
                () -> new EntityNotFoundException("Can't find category by id " + id));
```

waspp01 committed:
7th commit:
- Added validation in updateCategory method in CategoryController
- replaced Constructor in Category
- optimized getById method in CategoryServiceImpl
